pkg/controller/direct/sql/normalize.go

```diff
@@ -290,6 +303,95 @@ func normalizeAuditLogBucketRef(ctx context.Context, kube client.Reader, obj *kr
 	}
 }
 
+func normalizeSourceSQLInstanceRef(ctx context.Context, kube client.Reader, obj *krm.SQLInstance) (string, error) {
+	if obj.Spec.CloneSource == nil {
+		return "", nil
+	}
+
+	sqlInstanceRef := obj.Spec.CloneSource.SQLInstanceRef
+
+	if sqlInstanceRef.External != "" && sqlInstanceRef.Name != "" {
+		return "", fmt.Errorf("cannot specify both spec.settings.cloneSource.sqlInstanceRef.external and spec.settings.cloneSource.sqlInstanceRef.name")
+	}
+
+	if sqlInstanceRef.External != "" {
+		return sqlInstanceRef.External, nil
+	} else if sqlInstanceRef.Name != "" {
+		if sqlInstanceRef.Namespace == "" {
+			sqlInstanceRef.Namespace = obj.Namespace
+		}
+
+		key := types.NamespacedName{
+			Namespace: sqlInstanceRef.Namespace,
+			Name:      sqlInstanceRef.Name,
+		}
+
+		sqlInstance := &unstructured.Unstructured{}
+		sqlInstance.SetGroupVersionKind(krm.SQLInstanceGVK)
+		if err := kube.Get(ctx, key, sqlInstance); err != nil {
+			if apierrors.IsNotFound(err) {
+				return "", k8s.NewReferenceNotFoundError(krm.SQLInstanceGVK, key)
+			}
+			return "", fmt.Errorf("error reading referenced SQLInstance %v: %w", key, err)
+		}
+
+		sqlInstanceName, err := refs.GetResourceID(sqlInstance)
+		if err != nil {
+			return "", err
+		}
+		return sqlInstanceName, nil
+	} else {
+		return "", fmt.Errorf("must specify either spec.settings.cloneSource.sqlInstanceRef.external or spec.settings.cloneSource.sqlInstanceRef.name")
+	}
+}
+
+func normalizeSourceSQLDatabaseRefs(ctx context.Context, kube client.Reader, obj *krm.SQLInstance) ([]string, error) {
+	if obj.Spec.CloneSource == nil || obj.Spec.CloneSource.SQLDatabaseRefs == nil {
+		return nil, nil
+	}
+
+	sqlDatabaseRefs := obj.Spec.CloneSource.SQLDatabaseRefs
+	sqlDatabases := make([]string, len(sqlDatabaseRefs))
+
+	for i, sqlDatabaseRef := range sqlDatabaseRefs {
```

Comment:
Not a blocker,  but I believe we also want to check and ensure the list contains all external values or name values, but not a mix of both.

Reply:
Why? I think it should be ok to mix

Reply:
@yuwenma proposed it and suggested it will help us avoid complicated edge cases. Yuwen, could you elaborate on it a bit?

Reply:
I think that it is not really complicated to implement this, and customer may want the functionality. There could be some databases managed by kcc, others managed externally, but customer might want to clone all of them.

Reply:
SG!

Personally, I like the flexibility of supporting a mix of both. Implementation-wise, it's fine to start with a stricter rule (don't allow the mix) and then loosen it (allow the mix) if we think this provides a better UX.

Reply:
I think it's more complicated to add a check to ensure that all references in the list are either external or kcc-managed. That requires more code. However, am not sure about the scenario you mentioned where it could cause "complicated edge cases". Do you remember the details?

Reply:
Thanks @maqiuyujoyce for raising this. 
I put the rationale and some thoughts [here](https://docs.google.com/document/d/1gdbFRW0nfny8j7C_bUDuGG2Bk3z_WaC_eYhoLbuppG0/edit?tab=t.0#heading=h.g21v7n6bvl3z). The no-mix is a recommendation ("shall" rather than "must", see [RFC2119](https://datatracker.ietf.org/doc/html/rfc2119#section-3)) . And that it is for GCP unique list.

Reply:
Just spoke w/ @justinsb about this. We are thinking it is overkill to use refs for these SQLDatabases anyway, since they are associated with the parent SQLInstanceRef. Decided to make this field just a list of strings of database names.